src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Datadog.Trace.Agent.MessagePack;
+using Datadog.Trace.HttpOverStreams;
+using Datadog.Trace.HttpOverStreams.HttpContent;
+
+namespace Datadog.Trace.Agent.Transports
+{
+    internal class HttpStreamRequest : IApiRequest
+    {
+        private readonly Uri _uri;
+        private readonly DatadogHttpClient _client;
+        private readonly IStreamFactory _streamFactory;
+        private readonly HttpHeaders _headers = new HttpHeaders();
+
+        public HttpStreamRequest(DatadogHttpClient client, Uri uri, IStreamFactory streamFactory)
+        {
+            _uri = uri;
+            _client = client;
+            _streamFactory = streamFactory;
+        }
+
+        public void AddHeader(string name, string value)
+        {
+            _headers.Add(name, value);
+        }
+
+        public async Task<IApiResponse> PostAsync(Span[][] traces, FormatterResolverWrapper formatterResolver)
+        {
+            using (var bidirectionalStream = _streamFactory.GetBidirectionalStream())
+            {
+                // buffer the entire contents for now so we can determine its size in bytes and avoid chunking.
+                // TODO: support chunked transfer encoding to avoid buffering the entire contents
+                var requestContentStream = new MemoryStream();
+                await CachedSerializer.Instance.SerializeAsync(requestContentStream, traces, formatterResolver).ConfigureAwait(false);
+                requestContentStream.Position = 0;
+
+                var content = new StreamContent(requestContentStream, requestContentStream.Length);
```

Comment:
Nit: I was thinking maybe for the future we can create or own `HttpContent` and avoid the MemoryStream allocation and write directly to the underlying stream.

Reply:
I am currently working no memory streams for the response and after that I will do so for the request.